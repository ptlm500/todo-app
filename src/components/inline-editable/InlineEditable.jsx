import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default function InlineEditable(props) {
  const {
    className,
    value,
    placeholder,
    children,
    onSubmit
  } = props;

  if (Array.isArray(children)) {
    throw new Error('[InlineEditable]: Multiple child components passed. InlineEditable only accepts one child', children);
  }

  const childRef = useRef();

  const [editing, setEditing] = useState(false);

  useEffect(() => {
    if (editing && childRef && childRef.current) {
      childRef.current.focus();
    }
  }, [editing, childRef]);

  const childWithProps = React.cloneElement(children, {
    ref: childRef,
    value,
    placeholder
  });

  function handleKeyDown(e) {
    if (e.key === 'Escape') {
      finishEdit();
    }
  }

  function finishEdit() {
    setEditing(false);
    onSubmit && onSubmit(value);
  }

  return (
    <div
      className={className ? `${className} inline-editable` : 'inline-editable'}
    >
      {
        editing ? (
          <div
            className='inline-editable__edit-container'
            onBlur={finishEdit}
            onKeyDown={handleKeyDown}
          >
            {childWithProps}
          </div>
        ) : (
          <div
            className='inline-editable__edit-trigger'
            onClick={() => setEditing(true)}
          >
            <Value>
              {value || placeholder || 'Provide a placeholder value'}
            </Value>
          </div>
        )
      }
    </div>
  );
}

const Value = styled.span`
  white-space: pre-wrap;
  word-break: break-word;
`;

InlineEditable.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  onSubmit: PropTypes.func
};
