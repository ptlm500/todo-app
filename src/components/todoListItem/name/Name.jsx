import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import InlineEditable from '../../inline-editable';
import Input from '../../input';
import { theme } from '../../../styles';


function Name(props) {
  const {
    className,
    updateName
  } = props;
  const nameInStore = props.name;
  const [name, setName] = useState(nameInStore);

  useEffect(() => {
    // Set name state if stored name changes
    setName(nameInStore);
  }, [nameInStore]);

  return (
    <InlineEditable
      className={className}
      value={name}
      placeholder={' '}
      onSubmit={updateName}
    >
      <Input
        onChange={e => setName(e.target.value)}
      />
    </InlineEditable>
  );
}

Name.propTypes = {
  className: PropTypes.string.isRequired,
  updateName: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired
};

export default styled(Name)`
  ${theme.headerFontSize};
  font-weight: 600;
  min-width: 100px;
`;
