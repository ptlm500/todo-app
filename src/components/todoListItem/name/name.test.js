import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Name from './Name';
import InlineEditable from '../../inline-editable';
import 'jest-styled-components';

const defaultProps = {
  name: 'Test name',
  updateName: jest.fn(),
  className: 'test-class'
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Name', () => {
  it('renders a Name', () => {
    let tree;
    act(() => {
      tree = renderer.create(<Name {...defaultProps}/>);
    });

    const inlineEditable = tree.root.findByType(InlineEditable);
    expect(inlineEditable.props.value).toBe(defaultProps.name);
    expect(inlineEditable.props.placeholder).toBe(' ');

    const newName = 'New name';

    inlineEditable.props.onSubmit(newName);
    expect(defaultProps.updateName).toHaveBeenCalledWith(newName);

    expect(tree).toMatchSnapshot();
  });

  it('updates internal state correctly', () => {
    let tree;
    act(() => {
      tree = renderer.create(<Name {...defaultProps}/>);
    });

    const inlineEditable = tree.root.findByType(InlineEditable);
    const input = inlineEditable.props.children;

    // Test internal state updates
    const editedName = 'Test name edit';
    act(() => {
      input.props.onChange({target: {value: editedName}});
    });

    expect(inlineEditable.props.value).toBe(editedName);

    // Test props override internal state updates
    const newProps = {
      ...defaultProps,
      name: 'New name'
    };

    act(() => {
      tree.update(<Name {...newProps}/>);
    });

    expect(inlineEditable.props.value).toBe(newProps.name);
  });
});
