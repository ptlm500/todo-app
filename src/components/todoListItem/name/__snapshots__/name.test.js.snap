// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name renders a Name 1`] = `
<div
  className="sc-AxirZ gyXZKu test-class inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      Test name
    </span>
  </div>
</div>
`;
